crates/bevy_math/src/sampling/mesh_sampling.rs

```diff
@@ -0,0 +1,66 @@
+//! Functionality related to random sampling from triangle meshes.
+
+use crate::{
+    primitives::{Measured2d, Triangle3d},
+    ShapeSample, Vec3,
+};
+use rand::Rng;
+use rand_distr::{Distribution, WeightedAliasIndex};
+use thiserror::Error;
+
+/// A [distribution] that caches data to allow fast sampling from a collection of triangles.
+/// Generally used through [`sample`] or [`sample_iter`].
+///
+/// [distribution]: Distribution
+/// [`sample`]: Distribution::sample
+/// [`sample_iter`]: Distribution::sample_iter
+///
+/// Example
+/// ```
+/// # use bevy_math::{Vec3, primitives::*};
+/// # use bevy_math::sampling::mesh_sampling::UniformMeshSampler;
+/// # use rand::{SeedableRng, rngs::StdRng, distributions::Distribution};
+/// let faces = Tetrahedron::default().faces();
+/// let sampler = UniformMeshSampler::try_new(faces).unwrap();
+/// let rng = StdRng::seed_from_u64(8765309);
+/// // 50 random points on the tetrahedron:
+/// let samples: Vec<Vec3> = sampler.sample_iter(rng).take(50).collect();
+/// ```
+pub struct UniformMeshSampler {
+    triangles: Vec<Triangle3d>,
+    face_distribution: WeightedAliasIndex<f32>,
+}
+
+impl Distribution<Vec3> for UniformMeshSampler {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec3 {
+        let face_index = self.face_distribution.sample(rng);
+        self.triangles[face_index].sample_interior(rng)
+    }
+}
+
+/// An error that indicates that a [`UniformMeshSampler`] could not be constructed
+/// because its input data had a total area of zero.
+#[derive(Debug, Error)]
+#[error("Failed to form distribution: provided triangles have zero area")]
+pub struct ZeroAreaMeshError;
+
+impl UniformMeshSampler {
+    /// Construct a new [`UniformMeshSampler`] from a list of triangles.
+    ///
+    /// Returns an error if the collection of triangles would have zero surface area.
+    pub fn try_new<T: IntoIterator<Item = Triangle3d>>(
+        triangles: T,
+    ) -> Result<Self, ZeroAreaMeshError> {
+        let triangles: Vec<Triangle3d> = triangles.into_iter().collect();
+        let areas = triangles.iter().map(Measured2d::area).collect();
+
+        let Ok(face_distribution) = WeightedAliasIndex::new(areas) else {
```

Comment:
One nitpick I'd note here is that we may theoretically return a misleading error.

from the docs of `WeightedAliasIndex::new()`:
```
Returns an error if:

The vector is empty.
The vector is longer than u32::MAX.
For any weight w: w < 0 or w > max where max = W::MAX / weights.len().
The sum of weights is zero.
```

The vector being empty and the sum being 0 are what we're accounting for.
The other cases would return a very confusing error though.
I don't think we're exceeding u32::MAX anytime soon though.
`w < 0` could maybe trigger with a 0-sized triangle whose area is slightly negative due to numerical error (though I doubt Vec3::length is returning such a value anytime soon.)
`w > max` could happen with a large mesh which simultaneously has many triangles. (Although again unlikely, I didn't do the math here.)

As I looked into it, I don't think any of these scenarios are likely, so I don't think it's actually worth it. (Unless someone disagrees with this conclusion.)

Reply:
I ended up just bubbling up the error, since I think that's a little bit better. (In cases where the other things actually happen, the end user can at least handle them.)